Add a script for parsing the "dining hours" page from reslife

rye opened the pull request:
See #4841. (Closes #4841.)

This PR adds a script that loads and processes the human-entered data on the [dining hours](https://wp.stolaf.edu/reslife/dining-hours/) page. This script _could_ be run automatically via a GitHub Action or something like that, if we wanted to.

(See the updated schedule file on this branch for reference of what this script outputs.)

rye committed:
stav-covid: Create an empty script for generating hours

In response to (GH) #4841, I decided to take a stab at parsing the hours
that are listed on the reslife site.

In the past, we have shied away from such automation because it's quite
hard to discern the proper source of truth.  BonApp has been known to do
quite frankly whatever they want, and the website has not had such a
table published.

In this case, however, Stav Hall now has hours published on a page. [1]
This is primarily because each dorm has a strict window within which
they are expected to get each meal.  These hours are (seemingly) updated
daily, which would produce an unreasonable maintenance burden.  This
script will be responsible for dumping the hours from the dining-hours
website into the 1-3-stav.yaml source file, which can then be parsed and
loaded client-side.

[1]: https://web.archive.org/web/*/https://wp.stolaf.edu/reslife/dining-hours/

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

rye committed:
stav-covid: Download the dining-hours site and parse it

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

rye committed:
stav-covid: Loop over the tables

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

rye committed:
stav-covid: Clean up imports slightly

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

rye committed:
stav-covid: Parse each table's prior h3-sib

We build a regular expression to match on text that looks like a date,
then take the first match of that, parse it using strptime, and try to
bring that into the current year in a smart way.  (First, start with the
date we parsed, then replace the year component of that with the current
year, and then add a year if that happens to be before the current date
by more than 180 days.)

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

rye committed:
stav-covid: Add+use a helper function to grab meal names from th's

Hunt through the first row of the thead and find all <th> elements.  Use
meal_name_from (basically get_text with a sanity check) to get out the
text inside of the th.

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

rye committed:
stav-covid: Add a loop over meal-dorm combos

This will not account for the weird cases like meal 0 (the dorm name)
and weird extra rows in the tbody, but this does print out the right
columns for each of the rows and such.

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

rye committed:
stav-covid: Skip the column containing the dorm name

Since there's no label for the tr's in the tbody, we should skip the
first column since it is treated as a label.

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

rye committed:
stav-covid: Process the timespecs

For now, just split into a list with two times in it.

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

rye committed:
stav-covid: Parse the timespec members to datetimes

Unfortunately, we need to convert to datetimes in order to get working
datetime arithmetic later on, so we pass in the date for context.

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

rye committed:
stav-covid: Adjust the log message

This is done to make it more clear what information we're processing.

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

rye committed:
stav-covid: Add lunch adjustment

If the meal is listed as Lunch, then it should open after 6am and close
before 6pm.  We check the cases where either hour is before 6am and add
12 hours in either case.

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

rye committed:
stav-covid: Add Dinner adjustment

Dinner always occurs after 12pm.  If either t_open or t_close is before
12pm, we should bump it forward 12 hours.

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

rye committed:
stav-covid: Add oddity handling for open > close issues

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

rye committed:
stav-covid: Add two warnings for potential logic errors

These don't seem to be fired on my local machine, at least, but are nice
to have floating around.

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

rye committed:
stav-covid: Generate the schedules

If a schedule doesn't exist, pre-fill it, and if it does, fill it in by
adding new entries to the hours: subkey.

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

rye committed:
stav-covid: Add two more sanity checks

Don't process empty rows, and don't process rows with differing numbers
of tds from the meal row.

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

rye committed:
stav-covid: Pack the schedules into a YAML-able document

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

rye committed:
stav-covid: Get a filename to write the output into

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

rye committed:
stav-covid: Finally, render to yaml

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

rye committed:
hours: Update the Stav Hours using the stav-covid script

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

rye committed:
stav-covid: Only output the first two letters of the day name

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

rye committed:
Re-render Stav Hours

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

rye committed:
stav-covid: Strip leading 0 from hour component of from and to

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

rye committed:
Re-render Stav Hall hours data

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

drewvolz commented on scripts/stav-covid.py:
```diff
@@ -0,0 +1,153 @@
+output = requests.get(url).text
+
+# Now, parse it.
+output = BeautifulSoup(output)
```
One nit, feel free to ignore. 


> GuessedAtParserWarning: No parser was explicitly specified, so I'm using the best available HTML parser for this system ("html.parser"). This usually isn't a problem, but if you run this code on another system, or in a different virtual environment, it may use a different parser and behave differently.
>
> The code that caused this warning is on line 19 of the file s.py. To get rid of this warning, pass the additional argument 'features="html.parser"' to the BeautifulSoup constructor.

```suggestion
output = BeautifulSoup(output, features="html.parser")
```

rye replied:
Hmm, on my system it actually picks lxml… is there a preference between the two?

hawkrives replied:
I would prefer lxml, but it requires another package. Maybe html.parser?

rye replied:
Yeah, this isn't performance-critical so I'll opt to go with the (batteries-included) Python version.

drewvolz replied:
- [BeautifulSoup's own list of differences](https://www.crummy.com/software/BeautifulSoup/bs4/doc/#differences-between-parsers)
- [Short breakdown from SO](https://stackoverflow.com/questions/25714417/beautiful-soup-and-table-scraping-lxml-vs-html-parser)

Honestly, feel free to choose lxml.

+- |  html.parser
--|--
Advantages | Batteries included, Decent speed, Lenient (as of Python 2.7.3 and 3.2.)
Disadvantages | Not very lenient (before Python 2.7.3 or 3.2.2)

+- |  lxml
--|--
Advantages | Very fast, Lenient
Disadvantages | External C dependency

+- | html5lib
--|--
Advantages | Extremely lenient, Parses pages the same way a web browser does, Creates valid HTML5
Disadvantages | Very slow, External Python dependency

rye committed:
Merge branch 'master' into stav-covid

drewvolz commented on scripts/stav-covid.py:
```diff
@@ -0,0 +1,153 @@
+			# Split the timespec on the timespec_re matches
+			times = [time.strip() for time in timespec_re.split(timespec)]
+
+			if len(times) != 2:
+				warnings.warn("timespec {} did not split nicely".format(timespec))
+				continue
```
Cool, very happy to see this handling here considering there is a case in the data that breaks the regex

> UserWarning: timespec 10:15-10-45 did not split nicely

What are your thoughts on compensating? Does it make our program too brittle if we try to catch this case?

I'm thinking about how `timespec_re` matches for delimiters, and how timeparse splits on colons. If we find a dash where a colon _should_ be (that sounds bad as I type this), we could make a swap?

Eh. My idea sounds super brittle. Perhaps the happy path leaves less room for errors and thorny edge cases of further typos breaking the flow.

rye replied:
I was thinking it might be smart to limit the number of splits via `maxsplit` on [`re.split`](https://docs.python.org/3/library/re.html#re.split), (then further splitting any invalid things on word boundaries and rejoining with a `:` before finally giving up) to catch stuff like that.

If you feel like I should, I totally could!

drewvolz replied:
That's a decent algorithm. I'm okay with it as-is, but I wonder what @hawkrives thinks?

hawkrives replied:
I'm good with either!

rye replied:
I think I'll probably just opt to expect the happy path. If anyone complains, we do have the valid excuse that what was entered into the table is _technically_ wrong, and hopefully on the next pass it should get fixed. 😉

drewvolz commented on scripts/stav-covid.py:
```diff
@@ -0,0 +1,153 @@
+				t_close += datetime.timedelta(hours=12)
+
+			# The open time should not be the same or even close to the close
+			# time, so we should wrap it around.  (We do this by bumping up the )
```
Bumping up the ____ (close time)?

```suggestion
			# time, so we should wrap it around.  (We do this by bumping up the 
			# close time)
```

drewvolz reviewed (commented):
This is looking very nice, @rye! Thanks a ton.

rye committed:
stav-covid: Fix missing text in comment

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>
Co-authored-by: Drew Volz <volz.drew@gmail.com>

Co-authored-by: Drew Volz <volz.drew@gmail.com>

rye committed:
stav-covid: Use Python's html.parser library explicitly

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

rye committed:
stav-covid: Change rendering workflow

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

rye committed:
stav-covid: Change population to use nested dicts

Now, the subscript levels are

(dorm):
  (meal)
  title: <meal title>
  hours: [<entries from before>]

And this seems to generate the right output.

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

rye committed:
stav-covid: Add example to help text for -o

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

rye committed:
Re-render building hours for the various dorms

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

hawkrives commented on scripts/stav-covid.py:
```diff
@@ -0,0 +1,167 @@
+	body = table.find("tbody")
+
+	# NOTE(rye): Should only have one thead
+	meals = [meal_name_from(th) for th in table.find("thead").find("tr").find_all("th")]
```
Do we want to `assert len(meals) == 1`, then?

rye replied:
If I cared much about error handling, certainly—that'd require doing a separate search for `table.find_all("thead")` instead; I could totally do that, though.

hawkrives reviewed (approved):
One thought

rye committed:
stav-covid: Just display the meal as the title of the schedule

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

rye committed:
stav-covid: Also include a message about data scrape time

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

rye committed:
Update the per-dorm Stav files

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

rye committed:
stav-covid: Fix Lunch and Dinner processing logic

Instead of adjusting _both_ the start _and_ end times when _either_ is
out of expected bounds, now just adjust the ones we expect.

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

rye committed:
Re-render the data

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

rye committed:
stav-covid: Add a lot more warnings

Now we check things like time components themselves being wrong (e.g.
jumping too far forward into the next day, etc.)

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

rye committed:
stav-covid: Shrink the notes string

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

rye committed:
Update per-dorm Stav Hall hours with new notes field

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>
